pkg/cloudprovider/providers/vsphere/vclib/diskmanagers/vmdm.go

```diff
@@ -0,0 +1,239 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package diskmanagers
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/types"
+	"golang.org/x/net/context"
+	"k8s.io/kubernetes/pkg/cloudprovider/providers/vsphere/vclib"
+)
+
+// vmDiskManager implements VirtualDiskProvider interface for creating volume using Virtual Machine Reconfigure approach
+type vmDiskManager struct {
+	diskPath      string
+	volumeOptions *vclib.VolumeOptions
+	vmOptions     *vclib.VMOptions
+}
+
+// Create implements Disk's Create interface
+// Contains implementation of VM based Provisioning to provision disk with SPBM Policy or VSANStorageProfileData
+func (vmdisk vmDiskManager) Create(ctx context.Context, datastore *vclib.Datastore) (err error) {
```

Comment:
In order to create the disk with storage policy, we use the vmdm.Create(). For this, we do the following steps:
1. Create a dummy VM.
2. Create and attach the virtual disk (persistent volume) to this VM by doing a ReconfigureVM operation on VC.
3. Detach the virtual disk from dummy VM.
4. Delete the dummy VM.

Since there are multiple steps involved we have handle error conditions for each of them accordingly.  In this case, it might happen the disk failed to create which will be in step2 (or) disk is already created but failed to detach the disk from dummy VM  as in step 3 or delete the dummy VM in step 4. If step 2 fails, kubernetes will attempt to call CreateVolume() again which will try to create and attach the disk to the previously created dummy VM. If step3 fails, kubernetes will attempt to call CreateVolume() which in this we check if the disk is already attached to dummy VM, if yes we detach the volume.